lib/recursive-methods.rb

```diff
@@ -1,49 +1,85 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) ['n' being the number of times factorial computation has to be implemented until n reaches 0]
+# Space complexity: 0(n) ['n' being the number of function calls on the stack]
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    raise ArgumentError if n < 0
+    return 1 if n == 0
+    return n * factorial(n - 1)
+
+end
+
+# Time complexity: O(n^2)
+# Space complexity: O(n^2)
+def reverse(s, i = 1)
+    return s if s == "" || s.length < 2
+    reversed = reverse(s[1..-1])
+    reversed += s[0]
+    return reversed
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n) -still has to iterate over every element in string
+# Space complexity: O(n) -stack can get very large if string is big
+def reverse_inplace_helper(s, start, finish)
+    if start < finish
+        s[start], s[finish] = s[finish], s[start]
+        return reverse_inplace_helper(s, start + 1, finish - 1)
+    else
+        return s
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+    return reverse_inplace_helper(s, 0, s.length - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+    return 0 if n <= 0
+    return 2 + bunny(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def nested(s)
```

Comment:
Due to the slice you're doing, this is O(n^2) for both time/space.

Otherwise 👍